Rearrange Packet Types section and references (#2203)
quicwg/base-drafts: draft-ietf-quic-transport.md

# Packets and Frames {#packets-frames}

QUIC endpoints communicate by exchanging packets. Packets have confidentiality
and integrity protection (see {{packet-protected}}) and are carried in UDP
datagrams (see {{packet-coalesce}}).

This version of QUIC uses the long packet header (see {{long-header}}) during

> connection establishment.  Packets that carry the long header are Initial

Reviewer: ```suggestion
 connection establishment.  Packets that start with the long header are Initial
```

Reviewer: Or just "with the long header" as you say below.

## Coalescing Packets {#packet-coalesce}

Initial ({{packet-initial}}), 0-RTT Protected ({{packet-0rtt}}), and Handshake
({{packet-handshake}}) packets contain a Length field, which determines the end
of the packet.  The Length field covers both the Packet Number and Payload
fields, both of which are confidentiality protected and initially of unknown
length. The length of the Payload field is learned once header protection is
removed.

Using the Length field, a sender can coalesce multiple QUIC packets into one UDP
datagram.  This can reduce the number of UDP datagrams needed to complete the
cryptographic handshake and starting sending data.  Receivers MUST be able to
process coalesced packets.

Coalescing packets in order of increasing encryption levels (Initial, 0-RTT,

> Handshake, 1-RTT) makes it more likely the receiver will be able to process all

Reviewer: 2 lines below: ", so it can only be the last packed included in a UDP datagram."

## Coalescing Packets {#packet-coalesce}

> Packets using the Extended Long Header ({{extended-long-header}}) contain a

Author: In this format, they're the subsections of the ELH section.  We'd probably have to be more explicit if we take the non-indented layout @martinthomson proposes.

## Coalescing Packets {#packet-coalesce}

Initial ({{packet-initial}}), 0-RTT ({{packet-0rtt}}), and Handshake

> ({{packet-handshake}}) packets contain a Length field, which determines the end

Reviewer 2: I think we might want to either rename the Length field, or at least clarify that it's in a different place based on the Long Header Type.

This enforces cryptographic separation between the data sent in the different
packet sequence number spaces.  Each packet number space starts at packet number

> 1.  Subsequent packets sent in the same packet number space MUST increase the

Author: Bah, my editor sometimes does this and I don't catch it.  Let me correct that.

> 1.  Subsequent packets sent in the same packet number space MUST increase the

Reviewer 3: Hah! 0 _is_ special :-)

Example pseudo-code for packet number decoding can be found in
{{sample-packet-number-decoding}}.

> ### Starting Packet Numbers

Reviewer 4: I don't think that we need this still, but absent analysis, this can be moved under the Initial packet section.

In this version of QUIC, the following packet types with the long header are
defined:

| Type | Name                          | Section                     |
|-----:|:------------------------------|:----------------------------|
|  0x0 | Initial                       | {{packet-initial}}          |
|  0x1 | 0-RTT                         | {{packet-0rtt}}             |
|  0x2 | Handshake                     | {{packet-handshake}}        |
|  0x3 | Retry                         | {{packet-retry}}            |
{: #long-packet-types title="Long Header Packet Types"}

The header form bit, connection ID lengths byte, Destination and Source
Connection ID fields, and Version fields of a long header packet are
version-independent. The other fields in the first byte are version-specific.
See {{QUIC-INVARIANTS}} for details on how packets from different versions of
QUIC are interpreted.

The interpretation of the fields and the payload are specific to a version and
packet type.  While type-specific semantics for this version are described in
the following sections, several long-header packets in this version of QUIC

> contain these additional fields:

Reviewer 4: Is there a picture to accompany this, or are you relying on the Initial/Handshake/0-RTT pictures?

Author: Replying on the other pictures.

  either 0 bytes in length or between 4 and 18 bytes.
  {{negotiating-connection-ids}} describes the use of this field in more detail.

The following packet types are defined:

| Type | Name                          | Section                     |
|-----:|:------------------------------|:----------------------------|
|  0x0 | Initial                       | {{packet-initial}}          |

> |  0x1 | 0-RTT Protected               | {{packet-0rtt}}        |

Reviewer 4: I think that this is right, but counting is hard.

```suggestion
|  0x1 | 0-RTT Protected               | {{packet-0rtt}}             |
```

Any data in CRYPTO frames is discarded - and no longer retransmitted - when
Initial keys are discarded.

### 0-RTT {#packet-0rtt}

A 0-RTT packet uses long headers with a type value of 0x1, followed by the
Length and Packet Number fields. The first byte contains the Reserved and Packet
Number Length bits.  It is used to carry "early" data from the client to the
server as part of the first flight, prior to handshake completion. As part of
the TLS handshake, the server can accept or reject this early data.

> See Sections 2.3 of {{!TLS13}} for a discussion of 0-RTT data and its

Reviewer 4: ```suggestion
See Section 2.3 of {{!TLS13}} for a discussion of 0-RTT data and its
```

limitations.

~~~
+-+-+-+-+-+-+-+-+

> |1|1| 0 |R R|P P|

Reviewer 4: ```suggestion
|1|1| 1 |R R|P P|
```

Token:

> : The value of the token that was previously provided in a Retry packet or

Author: You're right, this fell in during the ELH combination.  So Initial packets *are* different because of the Token fields, which means that only 0-RTT and Handshake have the same ELH format.  Are we sure it's worth not simply having an explicit list of fields for those two?  That makes me more inclined to just revert the ELH change entirely and make each long-header packet type stand alone.

  NEW_TOKEN frame.

Length:

: The length of the remainder of the packet (that is, the Packet Number and
  Payload fields) in bytes, encoded as a variable-length integer
  ({{integer-encoding}}).

Packet Number:

: The packet number field is 1 to 4 bytes long. The packet number has
  confidentiality protection separate from packet protection, as described in
  Section 5.4 of {{QUIC-TLS}}. The length of the packet number field is encoded
  in the plaintext packet number. See {{packet-encoding}} for details.

Payload:

: The payload of the packet.

> ### Initial Packet {#packet-initial}

Reviewer 4: I think that we should use h2 for this.  The section is a mix of "here is a common format" and "here is a specific instance of a common format", but I think that's tolerable.  We would then have something like:

```
17. Packet Formats
    17.1. Packet Number Encoding and Decoding
    17.2. Long Header Packets <-- note the extra 's'
    17.3. Version Negotiation Packet
    17.4. Retry Packet
    17.5. Extended Long Header Packets
    17.6. Initial Packet
        17.6.1. Abandoning Initial Packets
        17.6.2. Starting Packet Numbers
    17.7. 0-RTT Protected Packet <-- extra 'Packet'
        17.7.1. 0-RTT Packet Numbers
    17.8. Handshake Packet
    17.9. Short Header Packets <-- extra 's'
```

Does that work?

Author: I understand the desire not to move subsections below h3, but the mix of "format" and "instance" really just doesn't flow for me.  But it does kind of match the rest of the document with the "grouped sections" motif, so maybe it's at least consistent.

(It's already contorted enough that VN is a long header packet, but not from this version and therefore absent from the type table.)

the connection.

### Handshake Packet {#packet-handshake}

A Handshake packet uses long headers with a type value of 0x2, followed by the
Length and Packet Number fields.  The first byte contains the Reserved and
Packet Number Length bits.  It is used to carry acknowledgments and
cryptographic handshake messages from the server and client.

~~~
+-+-+-+-+-+-+-+-+

> |1|1| 0 |R R|P P|

Reviewer 4: ```suggestion
|1|1| 2 |R R|P P|
```

Initial keys are discarded.

### 0-RTT Protected {#packet-0rtt}

A 0-RTT Protected packet uses extended long headers with a type value of 0x1.
It is used to carry "early" data from the client to the server as part of the
first flight, prior to handshake completion.  As part of the TLS handshake, the
server can accept or reject this early data.

See Sections 2.3 of {{!TLS13}} for a discussion of 0-RTT data and its

> limitations.

Reviewer 4: You should move the 0-RTT packet numbers section to here and remove the TODO.

Handshake protection keys are discarded.

> ## Short Header Packet {#short-header}

Reviewer 4: I'm assuming that the diff here looks terrible because it's moved text, not changed text.

Author: Yep.

> ~~~~~

Reviewer 4: ```suggestion
~~~
```
